sources/HUBViewControllerImplementation.m

```diff
@@ -1314,6 +1337,105 @@ - (void)removeComponentWrapperFromLookupTables:(nullable HUBComponentWrapper *)c
+                                                                                  scrollPosition:scrollPosition
+                                                                                  viewController:self];
+
+    CGRect const initialContentRect = [self contentRectForScrollView:self.collectionView];
+    [self.scrollHandler scrollingWillStartInViewController:self currentContentRect:initialContentRect];
+
+    __weak HUBViewControllerImplementation *weakSelf = self;
+    void (^completionWrapper)() = ^{
+        HUBViewControllerImplementation *strongSelf = weakSelf;
+        CGRect const destinationContentRect = [strongSelf contentRectForScrollView:strongSelf.collectionView];
+        [strongSelf.scrollHandler scrollingDidEndInViewController:strongSelf currentContentRect:destinationContentRect];
+        completionHandler();
+    };
+    
+    // If the component is already visible, the completion handler can be called instantly.
+    if ([self.collectionView.indexPathsForVisibleItems containsObject:rootIndexPath]) {
+        [self.collectionView setContentOffset:contentOffset animated:animated];
+        completionWrapper();
+    // If the scrolling is animated, the animation has to end before the new component can be retrieved.
+    } else if (animated) {
+        self.pendingScrollAnimationCallback = completionWrapper;
+        [self.collectionView setContentOffset:contentOffset animated:animated];
+    // If there's no animations, the UICollectionView will still defer updating the content offset to the next runloop cycle.
+    } else {
+        [self.collectionView setContentOffset:contentOffset animated:animated];
+        [[NSRunLoop currentRunLoop] runUntilDate:[NSDate date]];
+        completionWrapper();
+    }
+}
+
+- (void)scrollToRemainingComponentsStartingAtPosition:(NSUInteger)indexPosition
+                                            indexPath:(NSIndexPath *)indexPath
+                                            component:(nullable HUBComponentWrapper *)componentWrapper
+                                       scrollPosition:(UICollectionViewScrollPosition)scrollPosition
+                                             animated:(BOOL)animated
+                                           completion:(void (^ _Nullable)())completionHandler
+{
+    NSUInteger const childIndex = [indexPath indexAtPosition:indexPosition];
+
+    NSParameterAssert(childIndex != NSNotFound);
+    if (indexPosition > 0) {
+        NSParameterAssert(childIndex <= componentWrapper.model.children.count);
+    }
+
+    __weak HUBViewControllerImplementation *weakSelf = self;
+    void (^stepCompletionHandler)() = ^{
+        HUBViewControllerImplementation *strongSelf = weakSelf;
+        // If the scrolling isn't animated we have to wait for a runloop cycle, as scrolling is deferred.
+        if (!animated) {
+            [[NSRunLoop currentRunLoop] runUntilDate:[NSDate date]];
```

Comment:
Won't this block the main thread for the remainder of the cycle? Wouldn't it be better to `dispatch_async()`?